derotation/derotate_by_line.py

```diff
@@ -0,0 +1,110 @@
+import copy
+
+import numpy as np
+import tqdm
+from scipy.ndimage import rotate
+
+
+def rotate_an_image_array_line_by_line(
+    image_stack: np.ndarray,
+    rot_deg_line: np.ndarray,
+) -> np.ndarray:
+    """Rotates the image stack line by line, using the rotation angles
+    provided.
+
+    Description of the algorithm:
+    - takes one line from the image stack
+    - creates a new image with only that line
+    - rotates the line by the given angle
+    - substitutes the line in the new image
+    - adds the new image to the rotated image stack
+
+    Edge cases and how they are handled:
+    - the rotation starts in the middle of the image -> the previous lines
+    are copied from the first frame
+    - the rotation ends in the middle of the image -> the remaining lines
+    are copied from the last frame
+
+    Parameters
+    ----------
+    image_stack : np.ndarray
+        The image stack to be rotated.
+    rot_deg_line : np.ndarray
+        The rotation angles by line.
+
+    Returns
+    -------
+    np.ndarray
+        The rotated image stack.
+    """
+
+    num_lines_per_frame = image_stack.shape[1]
+    rotated_image_stack = copy.deepcopy(image_stack)
+    previous_image_completed = True
+    rotation_completed = True
+
+    min_value_img = np.min(image_stack)
+
+    for i, rotation in tqdm.tqdm(
+        enumerate(rot_deg_line), total=len(rot_deg_line)
+    ):
+        line_counter = i % num_lines_per_frame
+        image_counter = i // num_lines_per_frame
+
+        is_rotating = np.absolute(rotation) > 0.00001
+        image_scanning_completed = line_counter == (num_lines_per_frame - 1)
+        if i == 0:
+            rotation_just_finished = False
+        else:
+            rotation_just_finished = not is_rotating and (
+                np.absolute(rot_deg_line[i - 1]) > np.absolute(rotation)
+            )
+
+        if is_rotating:
+            if rotation_completed and (line_counter != 0):
+                # when starting a new rotation in the middle of the image
+                rotated_filled_image = (
+                    np.ones_like(image_stack[image_counter]) * min_value_img
+                )  # non sampled pixels are set to the min val of the image
```

Comment:
Thanks for the feedback Niko. Having this dark pixel was very useful in debugging because it helpted me see clearly when and how the derotation of the analysis made sense. 

I have then changed the code and now I am using the $F_0$ val of the luminance distribution, the same that should then be calculated by suite2p. 

What do you think? I am unsure of filling-in with interpolation. Maybe I can leave the choice to the user and let them decide if they want $F_0$ or the interpolation

Reply:
I think F0 is fine for now. You may add an option for interpolation in a later PR.